### PR TITLE
Strictly typed component decorator support

### PR DESCRIPTION
**Pull request by cmcrothers**
I have updated the types with an additional generic type so that standard connect calls are not aware of the void call.

I have also updated the tests to be more strict as the exisiting test did not attempt to render the component under test.

Addresses #9951

### CODE REVIEW

**Comment by seansfkelley** [comment]
@sheetalkamat hey, I was just about to review this. Why did you merge without approval?

File: react-redux/react-redux-tests.tsx [hunk 1]
```diff
@@ -34,14 +40,23 @@ class Counter extends Component<any, any> {
...
+}
 function mapStateToProps(state: CounterState) {
     return {
         value: state.counter
     };
 }
 
 // Which action creators does it want to receive by props?
-function mapDispatchToProps(dispatch: Dispatch<CounterState>) {
+function mapDispatchToProps(dispatch: Dispatch<Action>) {
```
Comment (seansfkelley) [inline react-redux/react-redux-tests.tsx]: This change is not correct -- `dispatch` is parameterized by the state it's relevant to, not by any action.

File: react-redux/react-redux.d.ts [hunk 2]
```diff
@@ -14,51 +14,72 @@ declare namespace ReactRedux {
   type Dispatch<S> = Redux.Dispatch<S>;
   type ActionCreator<A> = Redux.ActionCreator<A>;
 
-  interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps>|StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+
+  interface ComponentConstructor<P, S> extends __React.ComponentLifecycle<P, S> {
```
Comment (seansfkelley) [inline react-redux/react-redux.d.ts]: What purpose does this class serve? It looks a lot like `Component`, but slightly stricter in the constructor. Also, `S` in `ComponentLifecycle` is taken to be "state"; usage of it as `context` is not correct.

File: react-redux/react-redux.d.ts [hunk 3]
```diff
@@ -14,51 +14,72 @@ declare namespace ReactRedux {
...
+  new(props?: P, context?: S) : void;
+  }
+
+  interface ComponentClassDecorator<TOriginalProps, TOwnProps> {
+  (component: ComponentConstructor<TOriginalProps, any> | ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+  }
+
+  interface ComponentDecorator<TOriginalProps, TOwnProps, TState> {
```
Comment (seansfkelley) [inline react-redux/react-redux.d.ts]: So I see that the magic here seems to be that there is a third type parameter. Does this imply that every usage of `connect` has to be parameterized to ensure that it typechecks correctly?

File: react-redux/react-redux.d.ts [hunk 4]
```diff
@@ -14,51 +14,72 @@ declare namespace ReactRedux {
...
-   *
-   * Can't use the above decorator because it would default the type to {}
-   */
+  * Decorator that infers the type from the original component
+  *
+  * Can't use the above decorator because it would default the type to {}
+  */
   export interface InferableComponentDecorator {
-    <P, TComponentConstruct extends (ComponentClass<P>|StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
+  <P, TComponentConstruct extends (ComponentClass<P> | StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
```
Comment (seansfkelley) [inline react-redux/react-redux.d.ts]: Can you undo these changes to indentation (and elsewhere in the file) to stay consistent?

**Comment by sheetalkamat** [comment]
@JProgrammer please create a new PR addressing the feedback.

**Comment by seansfkelley** [comment]
@sheetalkamat Can we just revert this and start again since it was so recent? Also, why did you merge it without approval?

**Comment by sheetalkamat** [comment]
@mhegazy  asked me to review the prs and merge them. This one looked ok to me hence merged.

**Comment by seansfkelley** [comment]
@mhegazy I applaud the desire to merge these faster, but we should still try to get community reviews from typings owners (within reason)!

@sheetalkamat thank you for reverting, I appreciate it!

**Comment by seansfkelley** [comment]
@JProgrammer sorry about the roller-coaster, want to try this again?

**Comment by fraserharris** [comment]
@JProgrammer yes, please try again!

**Comment by zakdances** [comment]
Thank you! Looking forward to the end of `Unable to resolve signature of class decorator when called as an expression.` errors.

**Comment by rockmacaca** [comment]
@seansfkelley I've been spending the past few days trying to tackle this issue, but I'm running into a couple of major roadblocks (correct me if I'm wrong on something):

From what I've gathered, `@connect` doesn't really function as a true class decorator and isn't usable by TypeScript in its current form. The wrapped result of the connect call doesn't actually extend the targeted class, it uses its own type definition. TypeScript views decorated classes under the assumption that the resulting class will pass an `instanceof` check on the target. This just simply isn't true with what react-redux is doing.

The Connect class returned is an entirely new class that wraps the targeted class, with its own unique set of props. This is the main block here is because the props that the Connect class says are allowed are different than those that the class it's wrapping are. This is what causes the `Types of property 'props' are incompatible.` message when using it as a decorator. TypeScript can't verify that the Connect class is indeed an `instanceof` the wrapped class because this offending property.

To work around this, we can make the type declarations for props less strict. I am currently aware of two. The first is to say
```typescript
interface StateProps {
  myState: MyState
}
interface DispatchProps { ... }
interface OwnProps { ... }

@connect<StateProps, DispatchProps, OwnProps>(..., ...)
export class MyComponent extends Component<any, ComponentState> { ... }
```
saying that any props are assignable on the component. This has the disadvantage that TS can't vaildate anything pertaining to props. The other (and I think better) way is to make every property in `StateProps` and `DispatchProps` optional:
```typescript
interface StateProps {
  myState?: MyState
}
interface DispatchProps { ... }
interface OwnProps { ... }

@connect<StateProps, DispatchProps, OwnProps>(..., ...)
export class MyComponent extends Component<OwnProps & StateProps & DispatchProps, ComponentState> { ... }
```
This way everything checks out, but TS can still complain if props in `OwnProps` are omitted in JSX.

**Comment by seansfkelley** [comment]
Have you taken a look at #9951? It outlines some of the issues as you have here (such as https://github.com/Microsoft/TypeScript/issues/4881).

As for partial types, that is an improvement, though it requires you to compromise on those types available to the class implementation. https://github.com/Microsoft/TypeScript/pull/12114 ought to help, if this is the path the typings take.

**Comment by rockmacaca** [comment]
Thanks for pointing me to those, glad I'm not going crazy. Is there any point in messing with this further until we have mapped types then?

**Comment by seansfkelley** [comment]
I took a hard look at this in 1.8 and 2.0 hasn't introduced anything that would help (I think), so probably not.
